File: billing_system/frontend/flutter_application/lib/models/bill.dart
import 'bill_item.dart';

/// Saved bill returned from the Flask API.
class Bill {
  final String billNumber;
  final String date;
  final int customerId;
  final String customerName;
  final String customerPhone;
  final String paymentType;
  final String salesType;
  final String through;
  final String area;
  final String priceList;
  final String remarks;
  final List<BillItem> items;
  final double subtotal;
  final double gstTotal;
  final double roundOff;
  final double grandTotal;
  final Map<String, double> gstBreakup;
  final int itemCount;

  const Bill({
    required this.billNumber,
    required this.date,
    required this.customerId,
    required this.customerName,
    this.customerPhone = '',
    required this.paymentType,
    this.salesType = 'Retail',
    this.through = '',
    this.area = '',
    this.priceList = 'Retail',
    this.remarks = '',
    required this.items,
    required this.subtotal,
    required this.gstTotal,
    required this.roundOff,
    required this.grandTotal,
    required this.gstBreakup,
    required this.itemCount,
  });

  factory Bill.fromJson(Map<String, dynamic> json) {
    final rawBreakup = (json['gst_breakup'] as Map<String, dynamic>?) ?? {};
    final gstBreakup = rawBreakup.map(
      (k, v) => MapEntry(k, (v as num).toDouble()),
    );

    return Bill(
      billNumber:    json['bill_number']   as String,
      date:          json['date']          as String,
      customerId:    json['customer_id']   as int,
      customerName:  json['customer_name'] as String,
      customerPhone: json['customer_phone'] as String? ?? '',
      paymentType:   json['payment_type']  as String? ?? 'Cash',
      salesType:     json['sales_type']    as String? ?? 'Retail',
      through:       json['through']       as String? ?? '',
      area:          json['area']          as String? ?? '',
      priceList:     json['price_list']    as String? ?? 'Retail',
      remarks:       json['remarks']       as String? ?? '',
      items: (json['items'] as List<dynamic>)
          .map((e) => BillItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      subtotal:      (json['subtotal']      as num?)?.toDouble() ?? 0.0,
      gstTotal:      (json['gst_total']     as num?)?.toDouble() ?? 0.0,
      roundOff:      (json['round_off']     as num?)?.toDouble() ?? 0.0,
      grandTotal:    (json['grand_total']   as num).toDouble(),
      gstBreakup:    gstBreakup,
      itemCount:     json['item_count']    as int? ??
                     (json['items'] as List<dynamic>?)?.length ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
        'bill_number': billNumber,
        'date': date,
        'customer_id': customerId,
        'customer_name': customerName,
        'customer_phone': customerPhone,
        'payment_type': paymentType,
        'sales_type': salesType,
        'through': through,
        'area': area,
        'price_list': priceList,
        'remarks': remarks,
        'items': items.map((e) => e.toJson()).toList(),
        'subtotal': subtotal,
        'gst_total': gstTotal,
        'round_off': roundOff,
        'grand_total': grandTotal,
        'gst_breakup': gstBreakup,
        'item_count': itemCount,
      };
}

File: billing_system/backend/models/bill.py
"""
Bill and BillItem models.
"""

from dataclasses import dataclass, field
from typing import List, Optional
from datetime import datetime


@dataclass
class BillItem:
    product_id: int
    product_name: str
    unit: str
    quantity: float
    rate: float          # sale price per unit (no GST)
    discount_percent: float = 0.0

    # ------------------------------------------------------------------
    # Computed properties
    # ------------------------------------------------------------------

    @property
    def gross_amount(self) -> float:
        return round(self.rate * self.quantity, 2)

    @property
    def discount_amount(self) -> float:
        return round(self.gross_amount * self.discount_percent / 100, 2)

    @property
    def total(self) -> float:
        return round(self.gross_amount - self.discount_amount, 2)

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "product_id": self.product_id,
            "product_name": self.product_name,
            "unit": self.unit,
            "quantity": self.quantity,
            "rate": self.rate,
            "discount_percent": self.discount_percent,
            "discount_amount": self.discount_amount,
            "total": self.total,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "BillItem":
        return cls(
            product_id=data["product_id"],
            product_name=data["product_name"],
            unit=data.get("unit", "PCS"),
            quantity=float(data["quantity"]),
            rate=float(data["rate"]),
            discount_percent=float(data.get("discount_percent", 0)),
        )


@dataclass
class Bill:
    bill_number: str
    date: str                       # ISO-8601 string
    customer_id: int
    customer_name: str
    customer_phone: str = ""        # for internal records only
    payment_type: str = "Cash"      # "Cash" | "Credit" | "UPI"
    items: List[BillItem] = field(default_factory=list)
    remarks: str = ""
    sales_type: str = "Retail"
    through: str = ""               # salesman / agent
    area: str = ""
    price_list: str = "Retail"

    # ------------------------------------------------------------------
    # Computed totals
    # ------------------------------------------------------------------

    @property
    def subtotal(self) -> float:
        return round(sum(item.gross_amount for item in self.items), 2)

    @property
    def grand_total(self) -> float:
        return round(sum(item.total for item in self.items), 2)

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "bill_number": self.bill_number,
            "date": self.date,
            "customer_id": self.customer_id,
            "customer_name": self.customer_name,
            "customer_phone": self.customer_phone,
            "payment_type": self.payment_type,
            "sales_type": self.sales_type,
            "through": self.through,
            "area": self.area,
            "price_list": self.price_list,
            "remarks": self.remarks,
            "items": [item.to_dict() for item in self.items],
            "subtotal": self.subtotal,
            "grand_total": self.grand_total,
            "item_count": len(self.items),
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Bill":
        items = [BillItem.from_dict(i) for i in data.get("items", [])]
        return cls(
            bill_number=data["bill_number"],
            date=data.get("date", datetime.now().isoformat()),
            customer_id=data["customer_id"],
            customer_name=data["customer_name"],
            customer_phone=data.get("customer_phone", ""),
            payment_type=data.get("payment_type", "Cash"),
            items=items,
            remarks=data.get("remarks", ""),
            sales_type=data.get("sales_type", "Retail"),
            through=data.get("through", ""),
            area=data.get("area", ""),
            price_list=data.get("price_list", "Retail"),
        )
